openedx-unsupported/edx-analytics-pipeline
# Adds tasks to create Problem Response reports with Course Block records

This change adds a set of tasks which read `problem_check` events from the tracking logs, collate the "latest problem response" data into a Hive table, and generate CSV reports from this data.

**Discussions**: This change is related to [edx-analytics-data-api PR #128](https://github.com/edx/edx-analytics-data-api/pull/128), which has been discussed extensively with the Analytics team.

**Dependencies**: None

**Sandbox URL**: TBD

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:
1. See [edx-analytics-data-api PR 128](https://github.com/edx/edx-analytics-data-api/pull/128) for analyticstack test setup instructions.  
   
   Notes:
   - You don't need the Analytics API setup, or the custom ElasticSearch.
   - When installing the analytics-pipeline, take care to install this PR's branch:
   
   ```
   export PIPELINE_REPO="https://github.com/open-craft/edx-analytics-pipeline.git"
   export PIPELINE_BRANCH="jill/latest-problem-response"
   ```
   - And run this `launch-task` instead of `ImportEnrollmentsIntoMysql` or `ModuleEngagementWorkflowTask`:
   
   ```
   ProblemResponseReportWorkflow \
           --interval-start 2016-01-01  \
           --local-scheduler --n-reduce-tasks 2 --remote-log-level DEBUG \
           --output-root /edx/var/reports/ \
           --marker hdfs://localhost:9000/reports/marker`date +%s` \
           --hive-overwrite
   ```
   - You'll need to create `/edx/var/reports`, and ensure the hadoop user can write to it.
   - Use the `--hive-overwrite` parameter to see data changes if tracking logs change.
2. To view the problem response data, query the hive table created by the task:
   
   ```
   sudo su hadoop
   hive
   hive> select * from problem_response_latest;
   ```
3. To see the CSV reports, check `/edx/var/reports`.  There should be one file created there per course.
4. You can add more problem attempts via the LMS, and see them reflected in the reports by re-running the task.
   
   This task does not depend on any other enrolment or user profile data, so you can also dump any tracking.log files into hdfs and re-run the task, e.g.
   
   ```
   sudo su hadoop
   hdfs dfs -put -f /edx/src/tracking.log-real-data-from-course hdfs://localhost:9000/data/tracking.log-REAL
   ```
5. Run unit tests:
   
   ```
   vagrant ssh
   cd /var/lib/analytics-tasks/analyticstack/repo
   source ../venv/bin/activate
   export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
   export WHEEL_PYVER=python2.7
   make coverage
   ```

**Reviewers**
- [ ] @brianhw
